### PR TITLE
Various fixes for sway

### PR DESCRIPTION
**Pull request by no92**

### CODE REVIEW

File: options/posix/generic/sched-stubs.cpp [hunk 1]
```diff
@@ -31,18 +31,30 @@ int sched_get_priority_max(int) {
...
-	__builtin_unreachable();
+int __mlibc_cpu_count(const cpu_set_t *set) {
+	size_t i = 0;
+	size_t bit = 0;
+	size_t count = 0;
+
+	const unsigned char *ptr = reinterpret_cast<const unsigned char *>(set);
+
+	for(i=0; i < sizeof(cpu_set_t); i++) {
+		for(bit = 0; bit < 8; bit++) {
```
Comment (ArsenArsen) [inline options/posix/generic/sched-stubs.cpp:51]: do we prefer `CHAR_BITS` or w/e?

**Review by ArsenArsen — CHANGES_REQUESTED** [review]
The Linux implementation is missing

File: options/posix/generic/sched-stubs.cpp [hunk 2]
```diff
@@ -31,18 +32,30 @@ int sched_get_priority_max(int) {
 }
 
 int sched_get_priority_min(int) {
```
Comment (Geertiebear) [inline options/posix/generic/sched-stubs.cpp:34]: Please add logging to stubs like this.

**Review by Geertiebear — CHANGES_REQUESTED** [review]

File: options/posix/generic/sched-stubs.cpp [hunk 3]
```diff
@@ -31,18 +32,31 @@ int sched_get_priority_max(int) {
...
+	const unsigned char *ptr = reinterpret_cast<const unsigned char *>(set);
+
+	for(i = 0; i < sizeof(cpu_set_t); i++) {
+		for(bit = 0; bit < CHAR_BIT; bit++) {
+			if((1 << bit) & ptr[i])
+				count++;
+		}
+	}
+
+	return count;
```
Comment (ArsenArsen) [inline options/posix/generic/sched-stubs.cpp:59]: this doesn't match our style guidelines (and is also a very weird way to implement a popcnt). why are the loop counters declared out of line and initialized twice?

File: options/posix/generic/sched-stubs.cpp [hunk 4]
```diff
@@ -31,18 +32,31 @@ int sched_get_priority_max(int) {
 }
 
 int sched_get_priority_min(int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+	mlibc::infoLogger() << "mlibc: sched_get_priority_min() returns ENOSYS" << frg::endlog;
+	errno = ENOSYS;
+	return -1;
```
Comment (ArsenArsen) [inline options/posix/generic/sched-stubs.cpp:37]: can this be a sysdep stub?

**Review by Dennisbonke — APPROVED** [review]
LGTM

File: options/posix/generic/sched-stubs.cpp [hunk 5]
```diff
@@ -31,9 +31,15 @@ int sched_get_priority_max(int) {
 	__builtin_unreachable();
 }
 
-int sched_get_priority_min(int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int sched_get_priority_min(int policy) {
+	int res = 0;
+
+	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_getaffinity, -1);
```
Comment (ArsenArsen) [inline options/posix/generic/sched-stubs.cpp:37]: checking wrong sysdep (I suggest cherry-picking [`cfd64b461e210d903209b075e30f0048720565d1`](https://github.com/managarm/mlibc/commit/cfd64b461e210d903209b075e30f0048720565d1))

File: options/posix/generic/pthread-stubs.cpp [hunk 6]
```diff
@@ -499,9 +499,15 @@ int pthread_getname_np(pthread_t, char *, size_t) {
 	__builtin_unreachable();
 }
 
-int pthread_setschedparam(pthread_t, int, const struct sched_param *) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int pthread_setschedparam(pthread_t thread, int policy, const struct sched_param *param) {
+	auto tcb = reinterpret_cast<Tcb*>(thread);
+
+	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_setschedparam, ENOSYS);
```
Comment (ArsenArsen) [inline options/posix/generic/pthread-stubs.cpp:505]: wrong return value? (this isn't necessarily against spec afaik, but I don't think we should give return a mlibc-specific meaning of the return value)

File: options/posix/generic/pthread-stubs.cpp [hunk 7]
```diff
@@ -510,9 +510,15 @@ int pthread_setschedparam(pthread_t thread, int policy, const struct sched_param
 	return 0;
 }
 
-int pthread_getschedparam(pthread_t, int *, struct sched_param *) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int pthread_getschedparam(pthread_t thread, int *policy, struct sched_param *param) {
+	auto tcb = reinterpret_cast<Tcb*>(thread);
+
+	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_getschedparam, ENOSYS);
```
Comment (ArsenArsen) [inline options/posix/generic/pthread-stubs.cpp:516]: similarly here

**Review by ArsenArsen — CHANGES_REQUESTED** [review]

File: options/posix/generic/pthread-stubs.cpp (hunk 6 above)
Comment (no92) [reply options/posix/generic/pthread-stubs.cpp:505]: > On success, these functions return 0; on error, they return a nonzero error number.

I don't see how we could do any different TBH.

Comment (ArsenArsen) [reply options/posix/generic/pthread-stubs.cpp:505]: -1?

Comment (ArsenArsen) [reply options/posix/generic/pthread-stubs.cpp:505]: oh, error number, I misread the manpage, sorry

**Comment by no92** [comment]
Addressed the comments raised and rebased onto current master.

**Review by Dennisbonke — APPROVED** [review]
LGTM

**Pull request merged**